src/coreComponents/constitutive/contact/ContactBase.hpp

```diff
@@ -132,28 +151,113 @@ class ContactBase : public ConstitutiveBase
 
+
+  /// Type of kernel wrapper for in-kernel update
+  using KernelWrapper = ContactBaseUpdates;
+
+  /**
+   * @brief Create an update kernel wrapper.
+   * @return the wrapper
+   */
+  KernelWrapper createKernelWrapper() const;
+
   /**
    * @struct Structure to hold scoped key names
    */
   struct viewKeyStruct : public ConstitutiveBase::viewKeyStruct
   {
     /// string/key for penalty stiffness
     static constexpr char const * penaltyStiffnessString() { return "penaltyStiffness"; }
+
+    /// string/key for shear stiffness
+    static constexpr char const * shearStiffnessString() { return "shearStiffness"; }
+
+    /// string/key for aperture tolerance
+    static constexpr char const * apertureToleranceString() { return "apertureTolerance"; }
+
+    /// string/key for aperture table name
+    static constexpr char const * apertureTableNameString() { return "apertureTableName"; }
   };
 
 protected:
+  virtual void postProcessInput() override;
+
+  virtual void initializePreSubGroups() override;
+
+  /**
+   * @brief Validate the values provided in the aperture table
+   * @param[in] apertureTable the effective aperture vs aperture table
+   */
+  void validateApertureTable( TableFunction const & apertureTable ) const;
+
 
   /// The value of penalty to penetration
   real64 m_penaltyStiffness;
 
+  real64 m_shearStiffness;
+
+  /// The aperture tolerance to avoid floating point errors in expressions involving aperture
+  real64 m_apertureTolerance;
+
+  /// The name of the aperture table, if any
+  string m_apertureTableName;
+
+  /// Pointer to the function that limits the model aperture to a physically admissible value.
+  TableFunction const * m_apertureTable;
 };
 
+GEOSX_HOST_DEVICE
+void ContactBaseUpdates::computeTraction( localIndex const k,
+                                          arraySlice1d< real64 const > const & oldDispJump,
+                                          arraySlice1d< real64 const > const & dispJump,
+                                          arraySlice1d< real64 > const & tractionVector,
+                                          arraySlice2d< real64 > const & dTractionVector_dJump ) const
+{
+  GEOSX_UNUSED_VAR( k, oldDispJump );
+
+  tractionVector[0] = dispJump[0] >= 0 ? 0.0 : m_penaltyStiffness * dispJump[0];
+  tractionVector[1] = 0.0;
+  tractionVector[2] = 0.0;
+
+  LvArray::forValuesInSlice( dTractionVector_dJump, []( real64 & val ){ val = 0.0; } );
+  dTractionVector_dJump( 0, 0 ) = dispJump[0] >=0 ? 0.0 : m_penaltyStiffness;
+}
+
+
+GEOSX_HOST_DEVICE
+void ContactBaseUpdates::addPressureToTraction( real64 const & pressure,
+                                                bool const isOpen,
+                                                arraySlice1d< real64 > const & tractionVector,
+                                                real64 & dTraction_dPressure ) const
+{
+  tractionVector[0] -= pressure;
+  dTraction_dPressure = isOpen ? -1.0 : 0.0;
```

Comment:
the value and derivative aren't consistent with one another. They must either both depend on `isOpen` or both  be independent of `isOpen`. Physically I don't think it should depend on `isOpen` at all.  If there is pressure in the fracture, it will exert a traction, and its derivative should just be `-1` always.